core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -632,8 +633,9 @@ private[spark] object Utils extends Logging {
       fs: FileSystem,
       conf: SparkConf,
       hadoopConf: Configuration,
-      fileOverwrite: Boolean): Unit = {
-    if (!targetDir.mkdir()) {
+      fileOverwrite: Boolean,
+      filename: String = ""): Unit = {
```

Comment:
I'd use `Option[String] = None`. The in L648 you can do `val targetFile = new File(targetDir, filename.getOrElse(innerPath.getName)`.